File: algo/113. Path Sum II.js
// @9 Divide and Conquer

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number[][]}
 */

var pathSum = function(root, sum) {
    // if it is the child of leaf, return false
    if (root == null) {
        return [];
    }
    
    // if it is the leaf, check the value
    if (root.left==null && root.right==null) {
        if (root.val == sum) {
            return [[root.val]];
        } else {
            return [];
        }
    }
    
    // split
    var leftRes = pathSum(root.left, sum-root.val);
    var rightRes = pathSum(root.right, sum-root.val);
    
    // merge
    var res = [];
    
    if (leftRes) {
        for (var i=0; i<leftRes.length; i++) {
            var tempRes = leftRes[i];
            tempRes.unshift(root.val);
            res.push(tempRes);
        }
    }
    
    if (rightRes) {
        for (var i=0; i<rightRes.length; i++) {
            var tempRes = rightRes[i];
            tempRes.unshift(root.val);
            res.push(tempRes);
        }
    }
    return res;
};

File: round2/113. Path Sum II.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<IList<int>> PathSum(TreeNode root, int sum) {
        var res = new List<IList<int>>();
        if (root == null) return res;
        if (root.left == null && root.right == null) {
            if (root.val == sum) {
                //Console.WriteLine(sum);
                res.Add(new List<int>(){root.val});
                return res;
            } else {
                return res;
            }
        }
        var leftRes = PathSum(root.left, sum-root.val);
        var rightRes = PathSum(root.right, sum-root.val);
        foreach (var lst in leftRes) {
            lst.Insert(0, root.val);
            res.Add(lst);
        }
        foreach (var lst in rightRes) {
            lst.Insert(0, root.val);
            res.Add(lst);
        }
        return res;
    }
}